tensorflow_addons/custom_ops/activations/cc/kernels/gelu_op.h

```diff
@@ -0,0 +1,145 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_ADDONS_GELU_OP_H_
+#define TENSORFLOW_ADDONS_GELU_OP_H_
+
+#define EIGEN_USE_THREADS
+
+#include "tensorflow/core/framework/numeric_op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+namespace tensorflow {
+namespace functor {
+
+// Functor used by GeluOp to do the computations.
+template <typename Device, typename T>
+struct Gelu {
+  // Computes Gelu activation.
+  //
+  // features: any shape.
+  // approximate: whether to enable approximation.
+  // activations: same shape as "features".
+  void operator()(const Device& d, typename TTypes<T>::ConstTensor features,
+                  bool approximate, typename TTypes<T>::Tensor activations) {
+    if (approximate) {
+      // y = 0.5 * x * (1 + tanh(sqrt(2 / pi) * (x + 0.44715 * x^3)))
+      activations.device(d) =
+          static_cast<T>(0.5) * features *
+          (static_cast<T>(1) +
+           (static_cast<T>(M_2_SQRTPI * M_SQRT1_2) *
+            (features + static_cast<T>(0.044715) * features.cube()))
+               .tanh());
+    } else {
+      // y = x * normcdf(x) = 0.5 * x * (1 + erf(x / sqrt(2)))
+      activations.device(d) =
+          static_cast<T>(0.5) * features *
+          (static_cast<T>(1) + (features * static_cast<T>(M_SQRT1_2)).erf());
+    }
+  }
+};
+
+// Functor used by GeluGradOp to do the computations.
+template <typename Device, typename T>
+struct GeluGrad {
+  // Computes GeluGrad backprops.
+  //
+  // gradients: gradients backpropagated to the Gelu op.
+  // features: either the inputs that were passed to the Gelu or, or its
+  //           outputs (using either one yields the same result here).
+  // approximate: whether to enable approximation.
+  // backprops: gradients to backpropagate to the Gelu inputs.
+  void operator()(const Device& d, typename TTypes<T>::ConstTensor gradients,
+                  typename TTypes<T>::ConstTensor features, bool approximate,
+                  typename TTypes<T>::Tensor backprops) {
+    if (approximate) {
+      const T kAlpha = static_cast<T>(M_2_SQRTPI * M_SQRT1_2);
+      const T kBeta = kAlpha * static_cast<T>(0.044715) * static_cast<T>(3);
+      const auto y =
```

Comment:
I see here that the output `y` is recalculated. It was already calculated in the forward pass.

How about having the output of the forward pass, to be passed as an additional input to the backward pass in order to save time recalculating? i.e., have `typename TTypes<T>::ConstTensor activations` passed to the function, and replace this line with:
```
const auto y = activations
```

Reply:
I wonder if this causes excessive memory occupation. Such as BERT model, `gelu` is called multiple times in forward pass with multiple huge tensors of `[batch_size, sequence_length, depth]`. If we do the caching mechanism as you stated, the operation will store such large tensor in *each* forward node. I'm afraid it'll harm the model throughput because of OOM.

Or if you have already had some benchmarking results?

Reply:
Hello, seems that
`y = tanh(sqrt(2 / pi) * (x + 0.044715 * x^3)))`, and
`gelu(x) = 0.5 * x * (1 + tanh(sqrt(2 / pi) * (x + 0.044715 * x^3)))`

Though we can compute `y` via `y = gelu(x) / (0.5 * x) - 1`, this will indeed lose some precision during division. So I tend to preserve the current implementation. What's your thoughts on this @mostafaelhoushi and @AakashKumarNain .

Reply:
The current implementation looks good to me.

Reply:
So I think further changes could be done in the future. Let's move on to Layer's subclass! Thanks for you guys' feedback again.